morpho-checkpoint-node/src/Morpho/Config/Types.hs

```diff
@@ -116,43 +118,144 @@ deriving instance AllBF Eq f (NodeConfiguration_ Bare) => Eq (NodeConfiguration_
+      ncSystemStart = Compose fromJSON,
+      ncSecurityParameter = Compose fromJSON,
+      ncStableLedgerDepth = Compose fromJSON,
+      ncLoggingSwitch = Compose fromJSON,
+      ncTraceOpts =
+        TraceOptions
+          { traceVerbosity = Compose fromJSON,
+            traceChainDB = Compose fromJSON,
+            traceChainSyncClient = Compose fromJSON,
+            traceChainSyncHeaderServer = Compose fromJSON,
+            traceChainSyncBlockServer = Compose fromJSON,
+            traceBlockFetchDecisions = Compose fromJSON,
+            traceBlockFetchClient = Compose fromJSON,
+            traceBlockFetchServer = Compose fromJSON,
+            traceTxInbound = Compose fromJSON,
+            traceTxOutbound = Compose fromJSON,
+            traceLocalTxSubmissionServer = Compose fromJSON,
+            traceMempool = Compose fromJSON,
+            traceForge = Compose fromJSON,
+            traceChainSyncProtocol = Compose fromJSON,
+            traceBlockFetchProtocol = Compose fromJSON,
+            traceBlockFetchProtocolSerialised = Compose fromJSON,
+            traceTxSubmissionProtocol = Compose fromJSON,
+            traceLocalChainSyncProtocol = Compose fromJSON,
+            traceLocalTxSubmissionProtocol = Compose fromJSON,
+            traceLocalStateQueryProtocol = Compose fromJSON,
+            traceIpSubscription = Compose fromJSON,
+            traceDnsSubscription = Compose fromJSON,
+            traceDnsResolver = Compose fromJSON,
+            traceErrorPolicy = Compose fromJSON,
+            traceMux = Compose fromJSON,
+            traceHandshake = Compose fromJSON,
+            traceLedgerState = Compose fromJSON,
+            tracePoWNodeRpc = Compose fromJSON,
+            traceTimeTravelError = Compose fromJSON
+          },
+      ncTimeslotLength = Compose fromJSON,
+      ncSnapshotsOnDisk = Compose fromJSON,
+      ncSnapshotInterval = Compose fromJSON,
+      ncPoWBlockFetchInterval = Compose fromJSON,
+      ncPoWNodeRpcUrl = Compose fromJSON,
+      ncPrometheusPort = Compose fromJSON,
+      ncCheckpointInterval = Compose fromJSON,
+      ncRequiredMajority = Compose fromJSON,
+      ncFedPubKeys = Compose fromJSON,
+      ncNodePrivKeyFile = Compose fromJSON,
+      ncTopologyFile = TopologyFile <$> Compose fromJSON,
+      ncDatabaseDir = DbFile <$> Compose fromJSON,
+      -- TODO: Remove
+      ncSocketFile = SocketFile <$> Compose fromJSON,
+      ncNodeHost = NodeHostAddress . readMaybe . T.unpack <$> Compose fromJSON,
+      ncNodePort = (fromIntegral :: Int -> PortNumber) <$> Compose fromJSON,
+      ncValidateDb = Compose fromJSON,
+      ncLogging = Compose fromJSON
+    }
 
-defaultConfiguration :: NodeConfigurationFunctor Maybe
-defaultConfiguration =
-  emptyConfiguration
+-- | The configuration defaults for each field. Gets overridden by CLI and
+-- config file
+configFieldDefault :: NodeConfigurationFunctor Maybe
+configFieldDefault =
+  (bpure Nothing)
```

Comment:
Would be cool if we could do something like this for `configFieldParser`, but alas. (I have no idea if that's even possible or how. Maybe some custom parser type?)

Reply:
I haven't tried a whole lot of things, but just using `bpure` doesn't work, since it can't know that all fields are `FromJSON`. I'll probably try to improve this more in the future.